packages/zambdas/src/ehr/radiology/launch-viewer/validation.ts

```
@@ -0,0 +1,89 @@
+import Oystehr from '@oystehr/sdk';
+import { ServiceRequest } from 'fhir/r4b';
+import { isValidUUID, Secrets } from 'utils';
+import { ValidatedInput } from '.';
+import { validateJsonBody, ZambdaInput } from '../../../shared';
+
+export const validateInput = async (input: ZambdaInput, oystehr: Oystehr): Promise<ValidatedInput> => {
+  const serviceRequest = await validateBody(input, oystehr);
+
+  const callerAccessToken = input.headers.Authorization.replace('Bearer ', '');
+  if (callerAccessToken == null) {
+    throw new Error('Caller access token is required');
+  }
+
+  return {
+    serviceRequest,
+    callerAccessToken,
+  };
+};
+
+const validateBody = async (input: ZambdaInput, oystehr: Oystehr): Promise<ServiceRequest> => {
+  const { serviceRequestId } = validateJsonBody(input);
+
+  if (!isValidUUID(serviceRequestId)) {
+    throw new Error('serviceRequestId is required and must be a uuid');
+  }
+
+  let serviceRequest: ServiceRequest;
+  try {
+    serviceRequest = await oystehr.fhir.get<ServiceRequest>({
+      resourceType: 'ServiceRequest',
+      id: serviceRequestId,
+    });
+  } catch (error) {
+    if (error instanceof Error) {
+      console.log('Error fetching ServiceRequest in validate body: ', error.message);
+    }
+    throw new Error('Error fetching ServiceRequest in validate body');
```

Do we need a catch block if it only re-throws the error? The original error may be more informative.

---

I actually did this intentionally because I don't want to vomit out of the service unfriendly FHIR API errors like unspecific, "Not Found" up to the frontend. Seems it could accidentally leak some sensitive info to the frontend if we throw whatever from here.

That said, I want to clean up all the error handling in here to make a differentiation between 4xx and 5xx errors that we throw, and swallow messages for 500-code errors so that the message only ever says "Internal Error". If that were in place, then I would happily let this just throw up whatever, as I would know it would be swallowed safely by a top level catch block before return. 

Hope it makes sense :)

---

Do you mean proxy the response status from oystehr 400/500? Sounds good. Regarding error handling, I've started wrapping parameter validation in try/catch and throwing 400 errors there.

---

The workflow I like to use is throwing Errors of a certain shape that even include their own status code. `{ statusCode: 400, errorCode: 283947, message: "email must be a valid email address."}`. Then there is a top level try-catch in the zambda handler. It does something standard in every zambda: 
- if we caught a well-formed error object, then return it. these have exactly right status code / error code / message.
- if we caught anything else, return 500 + Internal Error message. This way we swallow any sensitive errors.

I'd like to bring that pattern to every zambda.